### PR TITLE
Add ring buffer op to NNVM and Relay

### PR DESCRIPTION
**Pull request by hcho3**
The ring buffer lets you cache last X outputs of a previous layer.
[buffer_op_design.pdf](https://github.com/neo-ai/tvm/files/3332610/buffer_op_design.pdf)

### CODE REVIEW

File: topi/python/topi/nn/buffer_op.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                                                data[i, j, k, l - buflen + data_size, m]),
+                               name='new_buffer')
+        if axis == 4:
+            return tvm.compute(buffer.shape,
+                               lambda i, j, k, l, m:
+                               tvm.if_then_else(m < buflen - data_size,
+                                                buffer[i, j, k, l, m + data_size],
+                                                data[i, j, k, l, m - buflen + data_size]),
+                               name='new_buffer')
+        assert False, "shouldn't get here"
```
Comment (kevinthesun) [inline topi/python/topi/nn/buffer_op.py:110]: More clear error msg?

**Comment by wweic** [comment]
@kevinthesun @zhiics could you take a look?
@hcho3 Could you rebase the branch?

**Comment by yongwww** [comment]
@hcho3 would you mind opening this pr in dmlc/tvm

**Comment by hcho3** [comment]
Will submit to dmlc/tvm instead.

**Pull request closed without merging**